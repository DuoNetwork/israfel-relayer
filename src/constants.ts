export const DB_ORDERS = 'orders';
export const DB_MAKER_ADDR = 'makerAddr';
export const DB_TIMESTAMP = 'timestamp';
export const DB_UPDATED_AT = 'updatedAt';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const RELAYER_HTTP_URL = 'http://localhost:3000/v0';
export const RELAYER_WS_URL = 'ws://localhost:8080';
export const MNEMONIC = 'concert load couple harbor equip island argue ramp clarify fence smart topic'
export const BASE_DERIVATION_PATH =  `44'/60'/0'/0`;
export const PENDING_HOURS = 24;

export const WS_CHANNEL_ORDERBOOK = 'orerbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
