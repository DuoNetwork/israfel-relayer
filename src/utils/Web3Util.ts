import {
	assetDataUtils,
	BigNumber,
	ContractWrappers,
	generatePseudoRandomSalt,
	orderHashUtils,
	RPCSubprovider,
	signatureUtils,
	SignedOrder,
	SignerType,
	Web3ProviderEngine
} from '0x.js';
import { schemas, SchemaValidator } from '@0xproject/json-schemas';
import { MnemonicWalletSubprovider } from '@0xproject/subproviders';
import { Web3Wrapper } from '@0xproject/web3-wrapper';
import * as CST from '../common/constants';
import { IRawOrder, IStringSignedOrder } from '../common/types';
import infura from '../keys/infura.json';
import util from './util';

export enum Wallet {
	None,
	Local,
	MetaMask,
	Ledger
}

export default class Web3Util {
	public contractWrappers: ContractWrappers;
	public web3Wrapper: Web3Wrapper;
	public wallet: Wallet = Wallet.None;
	public accountIndex: number = 0;

	constructor(window: any, live: boolean, mnemonic: string) {
		if (window && typeof window.web3 !== 'undefined') {
			this.web3Wrapper = new Web3Wrapper(window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else {
			const pe = new Web3ProviderEngine();
			pe.addProvider(
				new RPCSubprovider(
					(live ? CST.PROVIDER_INFURA_MAIN : CST.PROVIDER_INFURA_KOVAN) +
						'/' +
						infura.token
				)
			);
			if (!window && mnemonic) {
				const mnemonicWallet = new MnemonicWalletSubprovider({
					mnemonic: mnemonic,
					baseDerivationPath: CST.BASE_DERIVATION_PATH
				});
				pe.addProvider(mnemonicWallet);
			}
			pe.start();
			this.web3Wrapper = new Web3Wrapper(pe);
			this.wallet = window ? Wallet.None : Wallet.Local;
		}

		this.contractWrappers = new ContractWrappers(this.web3Wrapper.getProvider(), {
			networkId: live ? CST.NETWORK_ID_MAIN : CST.NETWORK_ID_KOVAN
		});
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		if (this.wallet !== Wallet.MetaMask) return;

		const store = (this.web3Wrapper.getProvider() as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public async getCurrentAddress(): Promise<string> {
		const accounts = await this.web3Wrapper.getAvailableAddressesAsync();
		return accounts[this.accountIndex] || CST.DUMMY_ADDR;
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3Wrapper.getNetworkIdAsync();
	}

	public async createRawOrder(
		userAddr: string,
		relayerAddr: string,
		makerAssetAddr: string,
		takerAssetAddr: string,
		makerAmt: number,
		takerAmt: number,
		expInSeconds: number
	): Promise<IRawOrder> {
		if (this.wallet !== Wallet.MetaMask) Promise.reject('cannot sign');
		const order = {
			senderAddress: relayerAddr.toLowerCase(),
			makerAddress: userAddr.toLowerCase(),
			takerAddress: relayerAddr.toLowerCase(),
			makerFee: new BigNumber(0),
			takerFee: new BigNumber(0),
			makerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(makerAmt), 18),
			takerAssetAmount: Web3Wrapper.toBaseUnitAmount(new BigNumber(takerAmt), 18),
			makerAssetData: assetDataUtils.encodeERC20AssetData(makerAssetAddr),
			takerAssetData: assetDataUtils.encodeERC20AssetData(takerAssetAddr),
			salt: generatePseudoRandomSalt(),
			exchangeAddress: this.contractWrappers.exchange.getContractAddress(),
			feeRecipientAddress: relayerAddr.toLowerCase(),
			expirationTimeSeconds: new BigNumber(expInSeconds)
		};

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const signature = await signatureUtils.ecSignOrderHashAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			order.makerAddress,
			SignerType.Metamask
		);
		return {
			orderHash: orderHash,
			signedOrder: { ...order, signature }
		};
	}

	public static getRandomFutureDateInSeconds = () => {
		return new BigNumber(Date.now() + CST.TEN_MINUTES_MS).div(CST.ONE_SECOND_MS).ceil();
	};

	public static stringToBN = (value: string): BigNumber => {
		return new BigNumber(value);
	};

	public static fromWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toUnitAmount(new BigNumber(value), 18).valueOf());
	};

	public static toWei = (value: BigNumber | string): number => {
		return Number(Web3Wrapper.toWei(new BigNumber(value)).valueOf());
	};

	public static assetDataToTokenName = (assetData: string): string => {
		const tokenAddr = assetDataUtils.decodeERC20AssetData(assetData).tokenAddress;
		return CST.TOKEN_MAPPING[tokenAddr];
	};

	public static getSideFromSignedOrder = (
		order: SignedOrder | IStringSignedOrder,
		pair: string
	): string => {
		return Web3Util.assetDataToTokenName(order.takerAssetData) === pair.split('-')[0]
			? CST.DB_BID
			: CST.DB_ASK;
	};

	public static getPriceFromSignedOrder = (order: IStringSignedOrder, side: string): number => {
		const isBid = side === CST.DB_BID;
		return util.round(
			Web3Util.stringToBN(isBid ? order.makerAssetAmount : order.takerAssetAmount)
				.div(isBid ? order.takerAssetAmount : order.makerAssetAmount)
				.valueOf()
		);
	};

	public async validateOrder(signedOrder: SignedOrder): Promise<string> {
		const { orderSchema } = schemas;
		const { signature, ...order } = signedOrder;
		const validator = new SchemaValidator();
		if (!validator.validate(order, orderSchema).valid) {
			util.logDebug('invalid schema ' + JSON.stringify(signedOrder));
			return '';
		}

		const orderHash = orderHashUtils.getOrderHashHex(order);
		const isValidSig = await signatureUtils.isValidSignatureAsync(
			this.web3Wrapper.getProvider(),
			orderHash,
			signature,
			order.makerAddress
		);
		if (!isValidSig) {
			util.logDebug('invalid signature ' + orderHash);
			return '';
		}

		return orderHash;
	}

	public getTokenAddressFromName(tokenName: string): string {
		switch (tokenName) {
			case CST.TOKEN_ZRX:
				return this.contractWrappers.exchange.getZRXTokenAddress();
			case CST.TOKEN_WETH:
				const ethTokenAddr = this.contractWrappers.etherToken.getContractAddressIfExists();
				if (!ethTokenAddr) {
					util.logInfo('no eth token address');
					return '';
				} else return ethTokenAddr;

			default:
				util.logInfo('no such token found');
				return '';
		}
	}

	public setAllUnlimitedAllowance(tokenAddr: string, addrs: string[]): Array<Promise<string>> {
		return addrs.map(address =>
			this.contractWrappers.erc20Token.setUnlimitedProxyAllowanceAsync(tokenAddr, address)
		);
	}

	public async setBaseQuoteAllowance(
		baseTokenAddr: string,
		quoteTokenAddr: string,
		addrs: string[]
	): Promise<void> {
		const responses = await Promise.all(
			this.setAllUnlimitedAllowance(quoteTokenAddr, addrs).concat(
				this.setAllUnlimitedAllowance(baseTokenAddr, addrs)
			)
		);
		await Promise.all(responses.map(tx => this.web3Wrapper.awaitTransactionSuccessAsync(tx)));
	}
}
