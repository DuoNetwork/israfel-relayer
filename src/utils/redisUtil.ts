import Redis from 'ioredis';
import * as CST from '../common/constants';
import {  IOrderBookSnapshotUpdate, IOrderQueueItem } from '../common/types';
import util from './util';

class RedisUtil {
	private redisPub: Redis.Redis | null = null;
	private redisSub: Redis.Redis | null = null;
	private handleOrderBookUpdate:
		| ((channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any)
		| null = null;
	private handleOrderUpdate:
		| ((channel: string, orderQueueItem: IOrderQueueItem) => any)
		| null = null;

	public init(redisKey: { host: string; password: string; servername: string }) {
		this.redisPub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
		this.redisSub.on('pmessage', (pattern, channel, message) =>
			this.onMessage(channel, message, pattern)
		);
	}

	public onMessage(channel: string, message: string, pattern: string = '') {
		util.logDebug(pattern + channel + message);
		const type = channel.split('|')[0];
		switch (type) {
			case CST.DB_ORDER_BOOKS:
				if (this.handleOrderBookUpdate)
					this.handleOrderBookUpdate(channel, JSON.parse(message));
				break;
			case CST.DB_ORDERS:
				if (this.handleOrderUpdate) this.handleOrderUpdate(channel, JSON.parse(message));
				break;
			default:
				break;
		}
	}

	public onOrderBookUpdate(
		handleOrderBookUpdate: (channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any
	) {
		this.handleOrderBookUpdate = handleOrderBookUpdate;
	}
	public onOrderUpdate(
		handleOrderUpdate: (channel: string, orderQueueItem: IOrderQueueItem) => any
	) {
		this.handleOrderUpdate = handleOrderUpdate;
	}

	public publish(channel: string, msg: string) {
		if (this.redisPub) return this.redisPub.publish(channel, msg);
		return Promise.resolve(0);
	}

	public increment(key: string) {
		if (this.redisPub) return this.redisPub.incr(key);
		return Promise.reject();
	}

	public set(key: string, value: string) {
		if (this.redisPub) return this.redisPub.set(key, value);
		return Promise.resolve('');
	}

	public get(key: string) {
		if (this.redisPub) return this.redisPub.get(key);
		return Promise.resolve('');
	}

	public hashSet(key: string, field: string, value: string) {
		if (this.redisPub) return this.redisPub.hset(key, field, value);
		return Promise.resolve(0);
	}

	public hashGet(key: string, field: string): Promise<string | null> {
		if (this.redisPub) return this.redisPub.hget(key, field);
		return Promise.resolve(null);
	}

	public async hashMultiGet(key: string, ...fields: string[]) {
		if (this.redisPub) {
			const values: Array<string | null> = await this.redisPub.hmget(key, ...fields);
			const output: { [field: string]: string | null } = {};
			fields.forEach((f, i) => (output[f] = values[i]));
			return output;
		}
		return Promise.resolve({});
	}

	public hashGetAll(key: string): Promise<any> {
		if (this.redisPub) return this.redisPub.hgetall(key);
		return Promise.resolve(null);
	}

	public hashDelete(key: string, field: string) {
		if (this.redisPub) return this.redisPub.hdel(key, field);
	}

	public subscribe(channel: string) {
		if (this.redisSub) this.redisSub.subscribe(channel);
	}

	public patternSubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.psubscribe(pattern);
	}

	public unsubscribe(channel: string) {
		if (this.redisSub) this.redisSub.unsubscribe(channel);
	}

	public patternUnsubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.punsubscribe(pattern);
	}

	public push(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.lpush(key, values);
	}

	public putBack(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.rpush(key, values);
	}

	public pop(key: string) {
		if (this.redisPub) return this.redisPub.rpop(key);
		return Promise.resolve('');
	}

	public multi() {
		if (this.redisPub) return this.redisPub.multi({ pipeline: false });
		return Promise.resolve('');
	}

	public exec() {
		if (this.redisPub) return this.redisPub.exec();
		return Promise.resolve('');
	}

	public getQueueLength(key: string) {
		if (this.redisPub) return this.redisPub.llen(key);
		return Promise.resolve(0);
	}
}

const redisUtil = new RedisUtil();
export default redisUtil;
