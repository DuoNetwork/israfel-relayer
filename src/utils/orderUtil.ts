import { signatureUtils, SignedOrder } from '0x.js';
import { schemas, SchemaValidator } from '@0xproject/json-schemas';
import Web3 from 'web3';
import * as CST from '../common/constants';
import {
	ICancelOrderQueueItem,
	ILiveOrder,
	INewOrderQueueItem,
	IOption,
	IStringSignedOrder,
	IUserOrder
} from '../common/types';
import infura from '../keys/infura.json';
import dynamoUtil from './dynamoUtil';
import redisUtil from './redisUtil';
import util from './util';

class OrderUtil {
	public getUserOrder(type: string, account: string, liveOrder: ILiveOrder): IUserOrder {
		return {
			account: account,
			pair: liveOrder.pair,
			type: type,
			status: CST.DB_CONFIRMED,
			orderHash: liveOrder.orderHash,
			price: liveOrder.price,
			amount: liveOrder.amount,
			side: liveOrder.side,
			sequence: liveOrder.currentSequence,
			updatedBy: CST.DB_ORDER_PROCESSOR
		};
	}

	public async addOrderToDB() {
		const res = await redisUtil.pop(`${CST.DB_ORDERS}|${CST.DB_ADD}`);

		if (res) {
			const orderQueueItem: INewOrderQueueItem = JSON.parse(res);
			try {
				await dynamoUtil.addRawOrder(orderQueueItem.rawOrder);
				await dynamoUtil.addLiveOrder(orderQueueItem.liveOrder);
				await dynamoUtil.addUserOrder(
					this.getUserOrder(
						CST.DB_ADD,
						orderQueueItem.rawOrder.signedOrder.makerAddress,
						orderQueueItem.liveOrder
					)
				);
			} catch (err) {
				redisUtil.putBack(`${CST.DB_ORDERS}|${CST.DB_ADD}`, res);
				return false;
			}

			return true;
		}
		return false;
	}

	public async cancelOrderInDB() {
		const res = await redisUtil.pop(`${CST.DB_ORDERS}|${CST.DB_CANCEL}`);
		if (res) {
			const orderQueueItem: ICancelOrderQueueItem = JSON.parse(res);
			// attention: needs to ensure atomicity of insertion
			try {
				await dynamoUtil.deleteRawOrderSignature(orderQueueItem.liveOrder.orderHash);
				await dynamoUtil.deleteLiveOrder(orderQueueItem.liveOrder);
				await dynamoUtil.addUserOrder(
					this.getUserOrder(
						CST.DB_CANCEL,
						orderQueueItem.account,
						orderQueueItem.liveOrder
					)
				);
			} catch (err) {
				redisUtil.putBack(`${CST.DB_ORDERS}|${CST.DB_CANCEL}`, res);
				return false;
			}

			return true;
		}
		return false;
	}

	public parseSignedOrder(order: IStringSignedOrder): SignedOrder {
		return {
			signature: order.signature,
			senderAddress: order.senderAddress,
			makerAddress: order.makerAddress,
			takerAddress: order.takerAddress,
			makerFee: util.stringToBN(order.makerFee),
			takerFee: util.stringToBN(order.takerFee),
			makerAssetAmount: util.stringToBN(order.makerAssetAmount),
			takerAssetAmount: util.stringToBN(order.takerAssetAmount),
			makerAssetData: order.makerAssetData,
			takerAssetData: order.takerAssetData,
			salt: util.stringToBN(order.salt),
			exchangeAddress: order.exchangeAddress,
			feeRecipientAddress: order.feeRecipientAddress,
			expirationTimeSeconds: util.stringToBN(order.expirationTimeSeconds)
		};
	}

	public async validateOrder(signedOrder: SignedOrder, orderHash: string): Promise<boolean> {
		const { orderSchema } = schemas;
		const { signature, ...rest } = signedOrder;
		const validator = new SchemaValidator();
		if (!validator.validate(rest, orderSchema).valid) {
			util.logDebug('invalid schema ' + orderHash);
			return false;
		}

		const isValidSig = await signatureUtils.isValidSignatureAsync(
			new Web3.providers.HttpProvider(CST.PROVIDER_INFURA_KOVAN + '/' + infura.token),
			orderHash,
			signature,
			rest.makerAddress
		);
		if (!isValidSig) {
			util.logDebug('invalid signature ' + orderHash);
			return false;
		}

		return true;
	}

	public async startProcessing(option: IOption) {
		dynamoUtil.updateStatus(
			option.type,
			await redisUtil.getQueueLength(`${CST.DB_ORDERS}|${option.type}`)
		);

		setInterval(
			async () =>
				dynamoUtil.updateStatus(
					option.type,
					await redisUtil.getQueueLength(`${CST.DB_ORDERS}|${option.type}`)
				),
			15000
		);

		const loop = () => {
			let promise: Promise<boolean>;
			switch (option.type) {
				case CST.DB_ADD:
					promise = this.addOrderToDB();
					break;
				case CST.DB_CANCEL:
					promise = this.cancelOrderInDB();
					break;
				default:
					throw new Error('Invalid type');
					break;
			}
			promise.then(result => {
				setTimeout(() => loop(), result ? 0 : 500);
			});
		};

		loop();
	}
}
const orderUtil = new OrderUtil();
export default orderUtil;
