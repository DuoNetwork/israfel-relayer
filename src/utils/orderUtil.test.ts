import * as CST from '../common/constants';
import dynamoUtil from './dynamoUtil';
import orderUtil from './orderUtil';
import redisUtil from './redisUtil';
import Web3Util from './Web3Util';

const signedOrder = {
	senderAddress: 'senderAddress',
	makerAddress: 'makerAddress',
	takerAddress: 'takerAddress',
	makerFee: '0',
	takerFee: '0',
	makerAssetAmount: '123',
	takerAssetAmount: '456',
	makerAssetData: 'makerAssetData',
	takerAssetData: 'takerAssetData',
	salt: '789',
	exchangeAddress: 'exchangeAddress',
	feeRecipientAddress: 'feeRecipientAddress',
	expirationTimeSeconds: '1234567890',
	signature: 'signature'
};

test('parseSignedOrder', () => expect(orderUtil.parseSignedOrder(signedOrder)).toMatchSnapshot());

test('constructNewLiveOrder', () => {
	Web3Util.getSideFromSignedOrder = jest.fn(() => CST.DB_BID);
	expect(orderUtil.constructNewLiveOrder(signedOrder, 'pair', '0xOrderHash')).toMatchSnapshot();
	Web3Util.getSideFromSignedOrder = jest.fn(() => CST.DB_ASK);
	expect(orderUtil.constructNewLiveOrder(signedOrder, 'pair', '0xOrderHash')).toMatchSnapshot();
});

const liveOrder = {
	account: '0xAccount',
	pair: 'pair',
	orderHash: '0xOrderHash',
	price: 0.123456789,
	amount: 456,
	side: CST.DB_BID,
	initialSequence: 1,
	currentSequence: 2
};

test('constructUserOrder', () => {
	expect(orderUtil.constructUserOrder(liveOrder, 'type', 'status', 'updatedBy'));
});

const addOrderQueueItem = {
	liveOrder: liveOrder,
	signedOrder: {
		senderAddress: 'senderAddress',
		makerAddress: 'makerAddress',
		takerAddress: 'takerAddress',
		makerFee: Web3Util.stringToBN('0'),
		takerFee: Web3Util.stringToBN('0'),
		makerAssetAmount: Web3Util.stringToBN('123'),
		takerAssetAmount: Web3Util.stringToBN('456'),
		makerAssetData: 'makerAssetData',
		takerAssetData: 'takerAssetData',
		salt: Web3Util.stringToBN('789'),
		exchangeAddress: 'exchangeAddress',
		feeRecipientAddress: 'feeRecipientAddress',
		expirationTimeSeconds: Web3Util.stringToBN('1234567890'),
		signature: 'signature'
	}
};

test('addOrderToDB no order', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(''));
	redisUtil.putBack = jest.fn();
	dynamoUtil.addRawOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.addOrderToDB();
	expect((dynamoUtil.addRawOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((dynamoUtil.addLiveOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((redisUtil.pop as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(false);
});

test('addOrderToDB', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(addOrderQueueItem)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.addRawOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.addOrderToDB();
	expect((dynamoUtil.addRawOrder as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((dynamoUtil.addLiveOrder as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(true);
});

test('addOrderToDB rawOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(addOrderQueueItem)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.addRawOrder = jest.fn(() => Promise.reject());
	dynamoUtil.addLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.addOrderToDB();
	expect((dynamoUtil.addLiveOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(1);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls[0][0]).toEqual(
		(redisUtil.pop as jest.Mock<Promise<boolean>>).mock.calls[0][0]
	);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls[0][1]).toEqual(
		JSON.stringify(addOrderQueueItem)
	);
	expect(isSuccess).toEqual(false);
});

test('addOrderToDB liveOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(addOrderQueueItem)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.addRawOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addLiveOrder = jest.fn(() => Promise.reject());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.addOrderToDB();
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(false);
});

test('addOrderToDB userOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(addOrderQueueItem)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.addRawOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.reject());
	const isSuccess = await orderUtil.addOrderToDB();
	expect(isSuccess).toEqual(false);
});

test('cancelOrderInDB no order', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(''));
	redisUtil.putBack = jest.fn();
	dynamoUtil.deleteRawOrderSignature = jest.fn(() => Promise.resolve());
	dynamoUtil.deleteLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.cancelOrderInDB();
	expect(
		(dynamoUtil.deleteRawOrderSignature as jest.Mock<Promise<boolean>>).mock.calls.length
	).toBe(0);
	expect((dynamoUtil.deleteLiveOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((redisUtil.pop as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(false);
});

test('cancelOrderInDB', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(liveOrder)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.deleteRawOrderSignature = jest.fn(() => Promise.resolve());
	dynamoUtil.deleteLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.cancelOrderInDB();
	expect(
		(dynamoUtil.deleteRawOrderSignature as jest.Mock<Promise<boolean>>).mock.calls
	).toMatchSnapshot();
	expect(
		(dynamoUtil.deleteLiveOrder as jest.Mock<Promise<boolean>>).mock.calls
	).toMatchSnapshot();
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls).toMatchSnapshot();
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(true);
});

test('cancelOrderInDB rawOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(liveOrder)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.deleteRawOrderSignature = jest.fn(() => Promise.reject());
	dynamoUtil.deleteLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.cancelOrderInDB();
	expect((dynamoUtil.deleteLiveOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(1);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls[0][0]).toEqual(
		(redisUtil.pop as jest.Mock<Promise<boolean>>).mock.calls[0][0]
	);
	expect((redisUtil.putBack as jest.Mock<Promise<boolean>>).mock.calls[0][1]).toEqual(
		JSON.stringify(liveOrder)
	);
	expect(isSuccess).toEqual(false);
});

test('cancelOrderInDB liveOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(liveOrder)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.deleteRawOrderSignature = jest.fn(() => Promise.resolve());
	dynamoUtil.deleteLiveOrder = jest.fn(() => Promise.reject());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.resolve());
	const isSuccess = await orderUtil.cancelOrderInDB();
	expect((dynamoUtil.addUserOrder as jest.Mock<Promise<boolean>>).mock.calls.length).toBe(0);
	expect(isSuccess).toEqual(false);
});

test('cancelOrderInDB userOrder failed', async () => {
	redisUtil.pop = jest.fn(() => Promise.resolve(JSON.stringify(liveOrder)));
	redisUtil.putBack = jest.fn();
	dynamoUtil.deleteRawOrderSignature = jest.fn(() => Promise.resolve());
	dynamoUtil.deleteLiveOrder = jest.fn(() => Promise.resolve());
	dynamoUtil.addUserOrder = jest.fn(() => Promise.reject());
	const isSuccess = await orderUtil.cancelOrderInDB();
	expect(isSuccess).toEqual(false);
});
