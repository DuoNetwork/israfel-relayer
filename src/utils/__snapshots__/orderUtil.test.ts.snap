// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addOrderToDB 1`] = `
Array [
  Array [
    Object {
      "orderHash": "0xOrderHash",
      "signedOrder": Object {
        "exchangeAddress": "exchangeAddress",
        "expirationTimeSeconds": "1234567890",
        "feeRecipientAddress": "feeRecipientAddress",
        "makerAddress": "makerAddress",
        "makerAssetAmount": "123",
        "makerAssetData": "makerAssetData",
        "makerFee": "0",
        "salt": "789",
        "senderAddress": "senderAddress",
        "signature": "signature",
        "takerAddress": "takerAddress",
        "takerAssetAmount": "456",
        "takerAssetData": "takerAssetData",
        "takerFee": "0",
      },
    },
  ],
]
`;

exports[`addOrderToDB 2`] = `
Array [
  Array [
    Object {
      "amount": 456,
      "currentSequence": 2,
      "initialSequence": 1,
      "orderHash": "0xOrderHash",
      "pair": "pair",
      "price": 0.123456789,
      "side": "bid",
    },
  ],
]
`;

exports[`addOrderToDB 3`] = `
Array [
  Array [
    Object {
      "account": "makerAddress",
      "amount": 456,
      "orderHash": "0xOrderHash",
      "pair": "pair",
      "price": 0.123456789,
      "sequence": 2,
      "side": "bid",
      "status": "confirmed",
      "type": "add",
      "updatedBy": "orderProcessor",
    },
  ],
]
`;

exports[`addOrderToDB no order 1`] = `
Array [
  Array [
    "orders|add",
  ],
]
`;

exports[`cancelOrderInDB 1`] = `
Array [
  Array [
    "0xOrderHash",
  ],
]
`;

exports[`cancelOrderInDB 2`] = `
Array [
  Array [
    Object {
      "amount": 456,
      "currentSequence": 2,
      "initialSequence": 1,
      "orderHash": "0xOrderHash",
      "pair": "pair",
      "price": 0.123456789,
      "side": "bid",
    },
  ],
]
`;

exports[`cancelOrderInDB 3`] = `
Array [
  Array [
    Object {
      "account": "0xAccount",
      "amount": 456,
      "orderHash": "0xOrderHash",
      "pair": "pair",
      "price": 0.123456789,
      "sequence": 2,
      "side": "bid",
      "status": "confirmed",
      "type": "cancel",
      "updatedBy": "orderProcessor",
    },
  ],
]
`;

exports[`cancelOrderInDB no order 1`] = `
Array [
  Array [
    "orders|cancel",
  ],
]
`;

exports[`parseSignedOrder 1`] = `
Object {
  "exchangeAddress": "exchangeAddress",
  "expirationTimeSeconds": "1234567890",
  "feeRecipientAddress": "feeRecipientAddress",
  "makerAddress": "makerAddress",
  "makerAssetAmount": "123",
  "makerAssetData": "makerAssetData",
  "makerFee": "0",
  "salt": "789",
  "senderAddress": "senderAddress",
  "signature": "signature",
  "takerAddress": "takerAddress",
  "takerAssetAmount": "456",
  "takerAssetData": "takerAssetData",
  "takerFee": "0",
}
`;
